### ISSUE TITLE
Nested stateless components do not hot reload

### ISSUE THREAD

**Issue by nickjanssen** [facebook/react-native#8465]
- Provide a **minimal code snippet** / [rnplay](https://rnplay.org/) example that reproduces the bug.

```
import React, { Component } from 'react';
import {
  AppRegistry,
  View,
  Text
} from 'react-native';

const ThisWillNotHotReload = () => {
  return (
    <Text>If change this text, it will not hot reload.</Text>
  )
}

const BugTest = () => {
  return (
    <View style={{
      flex: 1,
      justifyContent: 'center',
      alignItems: 'center'
    }}>
      <ThisWillNotHotReload />
    </View>
  )
}

AppRegistry.registerComponent('BugTest', () => BugTest);
```
- Provide **screenshots** where appropriate

Stateless components:
![hotreloadnotworking](https://cloud.githubusercontent.com/assets/2780160/16405653/6259a346-3d44-11e6-8649-431bbc74496a.gif)

Same but with a class:
![hotreloadworking](https://cloud.githubusercontent.com/assets/2780160/16405707/b4e83b22-3d44-11e6-8c07-c24951e16228.gif)
- What's the **version** of React Native you're using?

`0.28.0`
- Does this occur on iOS, Android or both?

both
- Are you using Mac, Linux or Windows?

Mac

**Comment by satya164** [comment]
cc @martinbigio

**Comment by olegdizus** [comment]
Have same issue

- React Native version: 0.29.0
- iOS
- Mac

**Comment by steida** [comment]
I think it should work for components placed in another file.
Maybe `AppRegistry.registerComponent('BugTest', () => BugTest);` is the problem.

**Comment by xgarrett** [comment]
- React Native 0.31.0
- Android
- Linux

I don't even need to nest the components. 

- components/App/index.android.js
```
import React from 'react';
import {
  Text,
  View
} from 'react-native';

const App = () => (
  <View>
    <Text>
      Welcome to React Native!
    </Text>
  </View>
);

export default App;
```

- index.android.js
```
import { AppRegistry } from 'react-native';
import App from './src/components/App';

AppRegistry.registerComponent('ReactNativeStarter', () => App);
```

**Comment by steida** [comment]
I think the trick is to have at least one ES6 component. Then it should work.

**Comment by khanghoang** [comment]
@steida could you elaborate you comment?

**Comment by steida** [comment]
https://github.com/este/este/blob/master/src/native/main.js#L37

**Comment by khanghoang** [comment]
@steida thanks. I tried but it doesn't work, i'm using react native ver 31.0.

**Comment by steida** [comment]
It works in Este, so get an inspiration from it.

**Comment by nbaldwin** [comment]
Same issue on 0.33, stateless functional component doesn't hot reload, works with ES6 class syntax.

- React Native version: 0.33.0
- iOS
- Mac

**Comment by zackify** [comment]
This still happens for me too. in 0.34.1

**Comment by Kitanotori** [comment]
Any fix on the horizon? Been using hot reloadable stateless components in React vanilla for some time now, and wouldn't want to switch back to stateful components if there is no actual need.

**Comment by jamielob** [comment]
Same issue on 0.35.0, iOS simulator and device, Mac.

**Comment by jcady** [comment]
Still an issue. 0.37.0 with iOS simulator on mac.

**Comment by sospedra** [comment]
+1 under  v0.37

**Comment by kylealwyn** [comment]
+1 on 0.40

**Comment by GeeWee** [comment]
+1

**Comment by sospedra** [comment]
Boomp! +1 on v0.40

**Comment by comountainclimber** [comment]
Will this issue be resolved anytime in the near future?

**Comment by jettandres** [comment]
+1! still existing on 0.44.0

**Comment by hramos** [comment]
If you're still running into this, the best approach right now would be to send a PR with a fix. Please refrain from adding metoo comments so that we can keep the discussion focused on a fix.

**Comment by bvic23** [comment]
Try my babel plugin https://github.com/bvic23/babel-plugin-functional-hmr until the issue is fixed.

**Comment by gaearon** [comment]
This has been fixed on master and should get into 0.61. https://github.com/facebook/react-native/issues/18899#issuecomment-506881769

If you have more cases that don't work please post reproducing examples in https://github.com/facebook/react-native/issues/18899.

**Comment by gaearon** [comment]
(The only caveat is that for Fast Refresh to work, you'll need to move your root component to a separate file)
